### PR TITLE
Advanced features for the MLDA step class

### PR DESCRIPTION
**Pull request by mikkelbue**
This PR integrates some advanced features to the MLDA stepper, along with performance improvements, extra examples and extra tests.

**Changes:**
- DEMetropolisZ step method is now the default base stepper for MLDA. 
  * This is implemented in the new class `DEMetropolisZMLDA`, which inherits from the original stepper and adds features necessary to run it as a base step method in MLDA. 
  * The `MLDA` step class and the `RecursiveDAProposal` class have been modified to support the new functionality.
  * MLDA notebooks have been rerun with this new default to reflect these changes.

- Added support for Adaptive Error Model (AEM), which recursively corrects coarse proposals, accoring to a Gaussian error model. For each adjacent model levels, it estimates the error (or _bias_) of the coarser model, and corrects its output according to the mean and covariance of this bias. This is broadly referred to as an Enhanced Error Model in the literature, and its efficiency has been demonstrated in the context of Delayed Acceptance MCMC numerous times[1]. 

  * Adaptive Error Models are implemented in the `__init__` and `astep` methods of the `MLDA` step class.
  * An extra method called `update_error_estimate` has been added to the `MLDA` class, to wrap the code that performs the update at each step. We have also added the class `RecursiveSampleMoments` that implements the iteratively constructed model mean/sigma update.
  * A new function, `delta_logp_inverse` has been added to `mlda.py`, which performs the same calculation as the existing `delta_logp`, but the order of the terms has been reversed. This is necessary for the adaptive error model to work.
  * AEM requires a custom Theano Op class. Details can be found in the docstring and the example notebook.

- An new Variance Reduction (VR) has been implemented, which allows the user to to specify a Quantity of Interest (QoI) before sampling, and calculates this QoI for each step at every level, as well as differences between adjacent levels. After the end of sampling, the user can employ the QoI differences to compute the QoI with reduced variance. This is a new method, which has been shown to work in practice, and the manuscript for a publication is under preparation. The logic is similar to that of [2], but does not require sampling independent coarse chains.

  * Similarly to the AEM, this has been implemented in  the `__init__` and `astep` methods of the `MLDA` step class, and an extra method called `update_vr_variables` that wraps the code has been added to the `MLDA` class.
  * Both base steppers have been updated to reflect this new functionality.
  * A new notebook that demonstrates VR on a linear regression example has been added.
   * VR requires a custom Theano Op class. Details can be found in the docstring and the example notebook.

- A new function called `subsample` has been added to `sampling.py`.
  * `subsample` is basically a stripped down version of `sample`, which skips a lot of checks and overhead, unnecessary for the MLDA sampler. This improves performance, because `subsample` is called many times recursively, inside the `RecursiveDAProposal`.
  * `RecursiveDAProposal` now caches the trace of successive calls to `subsample`. This avoid creating and destroying traces, and further reduces the overhead.

- Adds new tests for AEM and VR, and modifies existing tests to also test for DEMetropolisZ as a base sampler.

- Some important notebooks have been excluded from this PR, since they rely on external code (i.e. FEniCS). This includes groundwater flow examples for [AEM](https://github.com/alan-turing-institute/pymc3/blob/mlda_all_notebooks/docs/source/notebooks/MLDA_adaptive_error_model.ipynb), [VR](https://github.com/alan-turing-institute/pymc3/blob/mlda_all_notebooks/docs/source/notebooks/MLDA_variance_reduction_groundwater.ipynb) and an additional [benchmarking notebook](https://github.com/alan-turing-institute/pymc3/blob/mlda_all_notebooks/docs/source/notebooks/MLDA_benchmarks_tuning.ipynb). We are happy to host these notebooks elsewhere and supply an extra notebook with links to those resources within this PR.

[1] Cui, T., Fox, C., & O’Sullivan, M. J. (2018). A posteriori stochastic correction of reduced models in delayed acceptance MCMC, with application to multiphase subsurface inverse problems. ArXiv:1809.03176 [Stat]. http://arxiv.org/abs/1809.03176
[2] Dodwell, T. J., Ketelsen, C., Scheichl, R., & Teckentrup, A. L. (2015). A Hierarchical Multilevel Markov Chain Monte Carlo Algorithm with Applications to Uncertainty Quantification in Subsurface Flow. SIAM/ASA Journal on Uncertainty Quantification, 3(1), 1075–1108. https://doi.org/10.1137/130915005

### CODE REVIEW

File: pymc3/sampling.py [hunk 1]
```diff
@@ -631,6 +632,51 @@ def sample(
         return trace
 
 
+def subsample(
```
Comment (brandonwillard) [inline pymc3/sampling.py]: Does this need to be an addition to the core sampling module?  If it's step method-specific, as the docstring says, then this wouldn't be the place for it.

File: pymc3/sampling.py [hunk 2]
```diff
@@ -631,6 +632,51 @@ def sample(
...
+    trace=None,
+    tune=0,
+    model=None,
+    random_seed=None,
+    callback=None,
+    **kwargs
+):
+    """A stripped down version of sample(), which is called only
+    by the RecursiveDAProposal (which is the proposal used in the MLDA stepper).
+    It uses some all necessary arguments and checks for the context in which it
```
Comment (brandonwillard) [inline pymc3/sampling.py]: Is it "some" or "all"?

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: pymc3/sampling.py (hunk 1 above)
Comment (gmingas) [reply pymc3/sampling.py]: Hi @brandonwillard No this does not need to be in the core sampling module, we just placed it there as it is related to `sample()`. We can place it in step_methods/mlda.py if this is a better place for it?

File: pymc3/sampling.py (hunk 2 above)
Comment (gmingas) [reply pymc3/sampling.py]: This should be "some", will correct.

File: pymc3/sampling.py (hunk 1 above)
Comment (mikkelbue) [reply pymc3/sampling.py]: @brandonwillard Fixed in [17a4f39 ](https://github.com/alan-turing-institute/pymc3/commit/06e43e9464c9541624c783cf046c09c2e42fbd5d)

File: pymc3/sampling.py (hunk 2 above)
Comment (mikkelbue) [reply pymc3/sampling.py]: @brandonwillard  Fixed in [17a4f39](https://github.com/alan-turing-institute/pymc3/commit/06e43e9464c9541624c783cf046c09c2e42fbd5d)

**Comment by gmingas** [comment]
Just added a change in the way that subchain sampling is done when variance reduction is activated in order for the theory to work out. The coarse chain generates a sequence of samples and the fine chain selects a random one from the sequence (instead of the last one as previously). Also, the latest commits add a utility function to extract quantity of interest estimates and standard error from the trace.

**Comment by twiecki** [comment]
Let us know when you want us to merge. As this mostly touches your own code I think you can just do the code review in your group.

**Comment by gmingas** [comment]
> Let us know when you want us to merge. As this mostly touches your own code I think you can just do the code review in your group.

Makes sense. On thing we wanted to get your opinion on is whether it is OK to add [this](https://github.com/pymc-devs/pymc3/blob/061ecca51f638eda8d5e34cce683098d8fe63f48/docs/source/notebooks/MLDA_introduction.ipynb) notebook that contains introductory comments on MLDA plus links to all the other notebooks (hosted in PyMC3 and externally). We think it will help people find all the resources more easily (especially the FEniCS-based notebooks hosted by us).

**Comment by MarcoGorelli** [comment]
Hi @gmingas - there's been some CI changes on `master`, do you want fetch and merge `upstream/master` to avoid having to fix more merge conflicts later on?

**Comment by gmingas** [comment]
> Hi @gmingas - there's been some CI changes on `master`, do you want fetch and merge `upstream/master` to avoid having to fix more merge conflicts later on?

@MarcoGorelli We will rebase to incorporate the changes.

**Comment by twiecki** [comment]
> Makes sense. On thing we wanted to get your opinion on is whether it is OK to add this notebook that contains introductory comments on MLDA plus links to all the other notebooks (hosted in PyMC3 and externally). We think it will help people find all the resources more easily (especially the FEniCS-based notebooks hosted by us).

Definitely.

**Comment by mikkelbue** [comment]
@brandonwillard I made the changes you requested. Could you have a look and marked it as solved, if you are happy with it?

@twiecki We have merged all updates, and the checks are now passing. We are ready to squash-merge.

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Thanks guys!

**Pull request merged**